frontend/controller/service-worker.js

```diff
@@ -13,23 +15,95 @@ sbp('sbp/selectors/register', {
-        navigator.serviceWorker.controller.postMessage({
-          type: 'set',
-          key: 'foo',
-          value: 'bar'
+
+      if (Notification.permission !== 'granted') {
+        // To be able to receive a notification, this setting must be turned on.
+        // But user can customise this setting later via 'user-settings' page.
+        await requestNotificationPermission(true)
+      }
+
+      const API_URL = sbp('okTurtles.data/get', 'API_URL')
+      const existingSubscription = await registration.pushManager.getSubscription()
+
+      if (existingSubscription) {
+        // If there is an existing subscription, no need to create a new one.
+        // But make sure server knows the subscription details too.
+        await fetch(`${API_URL}/push/subscribe`, { method: 'POST', body: JSON.stringify(existingSubscription.toJSON()) })
+        return
+      } else {
+        // Create a new push subscription
+
+        // 1. Get the public key from the server
+        const PUBLIC_VAPID_KEY = await fetch(`${API_URL}/push/publickey`)
+          .then(handleFetchResult('text'))
+
+        // 2. Add a new subscription to pushManager using it.
+        const subscription = await registration.pushManager.subscribe({
+          userVisibleOnly: true,
+          applicationServerKey: urlBase64ToUint8Array(PUBLIC_VAPID_KEY)
         })
-      }, 1000)
+
+        // 3. Send the subscription details to the server. (server needs it to send the push notification)
+        await fetch(`${API_URL}/push/subscribe`, { method: 'POST', body: JSON.stringify(subscription.toJSON()) })
+
+        // 4. Send the test notification to confirm it works as expected. (Just a demo for now, but can be removed after development)
+        const testNotification = {
+          title: 'Service worker installed.',
+          body: 'You can now receive various push notifications from the Group Income app!',
+          endpoint: subscription.endpoint
+        }
+
+        await fetch(`${API_URL}/push/send`, { method: 'POST', body: JSON.stringify(testNotification) })
+      }
     } catch (e) {
-      // TODO: this
       console.error('error setting up service worker:', e)
     }
+  },
+  'service-worker/send-push': async function (payload) {
+    if (!('serviceWorker' in navigator)) { return }
+
+    const swRegistration = await navigator.serviceWorker.ready
+
+    if (!swRegistration) {
+      console.error('No service-worker registration found!')
+      return
+    }
+
+    const pushSubscription = await swRegistration.pushManager.getSubscription()
+
+    if (pushSubscription) {
+      const API_URL = sbp('okTurtles.data/get', 'API_URL')
+
+      await fetch(
+        `${API_URL}/push/send`,
+        { method: 'POST', body: JSON.stringify({ ...payload, endpoint: pushSubscription.endpoint }) }
+      )
+    } else {
+      console.error('No existing subscription found!')
+    }
   }
 })
+
+// helper method
+
+function urlBase64ToUint8Array (base64String) {
```

Comment:
Shorter version:

```
new Uint8Array(atob('AAAAAA'.replace(/-/g, '+').replace(/_/g, '/')).split('').map((c) => c.charCodeAt(0)))
```

A few notes:

1. I don't believe the padding is needed
2. I'd just all `atob` (as above). If you must refer to the global object, I'd use `self` or `globalThis` over `window`, as these are more generic.

Reply:
@corrideat good explanation but I would rather leave it as is if they both eventually do the same job.
FYI, This method is originally from [here](https://gist.github.com/Klerith/80abd742d726dd587f4bd5d6a0ab26b6), which was referenced in multiple articles that feature the `web-push` npm package. The fact that it was mentioned in those popular articles made me decide it's safe to assume it's trusted by the Javascript community.

Reply:
Re the point **2.**, At the moment this method is only used in browser context. So using `window` is fine.